File: sdk-python/examples/prices/error_handling.py
"""Run with: python examples/prices/error_handling.py"""

from __future__ import annotations

from threecommon import (
    AuthError,
    NotFoundError,
    RateLimitError,
    ThreeCommon,
    ValidationError,
)
from threecommon.prices import CreateBody


def main() -> None:
    with ThreeCommon(api_key="3co_your_api_key_here") as client:
        try:
            # `recurring` is required when type is `recurring`; omitting it
            # triggers a 400 validation error.
            client.prices.create(
                CreateBody(
                    product_id="prod_replace_with_real_id",
                    type="recurring",
                    currency="USD",
                    unit_amount=1500,
                )
            )
        except ValidationError as e:
            print(f"validation: {e.message}")
        except NotFoundError:
            print("product not found")
        except AuthError as e:
            print(f"auth failed: check your API key — code={e.code}")
        except RateLimitError as e:
            wait = e.retry_after_seconds or 30
            print(f"rate limited; waiting {wait}s before retry")


if __name__ == "__main__":
    main()

File: sdk-node/examples/prices/error-handling.ts
/**
 * Demonstrate the typed error tree on the prices surface. Each subclass
 * extends `ThreeCommonError`; branch with `instanceof`.
 *
 * Run:
 *   npx tsx examples/prices/error-handling.ts
 */

import {
  ThreeCommon,
  ThreeCommonAuthError,
  ThreeCommonNotFoundError,
  ThreeCommonRateLimitError,
  ThreeCommonValidationError,
} from '@3common/sdk'

const client = new ThreeCommon({
  apiKey: '3co_your_api_key_here',
})

try {
  // `recurring` is required when type is `recurring` — omitting it triggers a
  // 400 validation error.
  await client.prices.create({
    productId: 'prod_replace_with_real_id',
    type: 'recurring',
    currency: 'USD',
    unitAmount: 1500,
  })
} catch (err) {
  if (err instanceof ThreeCommonValidationError) {
    console.warn(`validation: ${err.message}`)
  } else if (err instanceof ThreeCommonNotFoundError) {
    console.warn('product not found')
  } else if (err instanceof ThreeCommonAuthError) {
    console.warn('bad or expired API key')
  } else if (err instanceof ThreeCommonRateLimitError) {
    const wait = err.retryAfterSeconds ?? 30
    console.warn(`rate limited; retry in ${String(wait)}s`)
  } else {
    throw err
  }
}
